aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -0,0 +1,366 @@
+
+#include <ATen/FunctionalTensorWrapper.h>
+
+#include <ATen/FunctionalInverses.h>
+#include <ATen/WrapDimUtils.h>
+#include <ATen/core/LegacyTypeDispatch.h>
+#include <c10/util/Exception.h>
+
+#include <c10/util/irange.h>
+
+namespace at {
+
+FunctionalTensorWrapper::FunctionalTensorWrapper(Tensor value)
+  : c10::TensorImpl(
+      c10::Storage(c10::make_intrusive<functionalization::FunctionalStorageImpl>(value.device(), value.numel(), value.dtype())),
+      c10::DispatchKeySet(DispatchKey::Functionalize),
+      value.dtype()
+    ),
+    value_(value),
+    // Note: "level" is a concept that we don't know how to compute in core.
+    // For now I'm retroactively setting this in functorch,
+    // but once Open Multiple Dispatch lands we should be able to calculate this in core.
+    level_(-1)
```

Comment:
NB: you wouldn't actually calculate this in core, instead, you'd do the lifting operation to turn functionalize into an open multiple dispatch. This is also pretty interesting because unlike Autograd/Conjugate lifting, functionalization involves its own wrapper, and I don't know if the design actually accounts for this case. cc @albanD